# Comparing two Linear Regression models for Predicting Student Academic Performance

## Project Objective

The objective of this project is to predict students' final grades (G3) using Linear Regression and compare two different feature sets.

### Research Questions

1. Can we accurately predict a student's final grade if previous grades (G1 and G2) are known?

2. Can demographic, behavioral, and academic characteristics predict final performance without using previous grades?

The performance of both models is evaluated using:

- Mean Absolute Error (MAE)
- Root Mean Squared Error (RMSE)
- R² Score

In [27]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

In [28]:
df = pd.read_csv("student-mat.csv")

In [29]:
df.head()
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   school      395 non-null    str  
 1   sex         395 non-null    str  
 2   age         395 non-null    int64
 3   address     395 non-null    str  
 4   famsize     395 non-null    str  
 5   Pstatus     395 non-null    str  
 6   Medu        395 non-null    int64
 7   Fedu        395 non-null    int64
 8   Mjob        395 non-null    str  
 9   Fjob        395 non-null    str  
 10  reason      395 non-null    str  
 11  guardian    395 non-null    str  
 12  traveltime  395 non-null    int64
 13  studytime   395 non-null    int64
 14  failures    395 non-null    int64
 15  schoolsup   395 non-null    str  
 16  famsup      395 non-null    str  
 17  paid        395 non-null    str  
 18  activities  395 non-null    str  
 19  nursery     395 non-null    str  
 20  higher      395 non-null    str  
 21  inte

In [30]:
def evaluate_model(name, y_test, predictions):

    mae = mean_absolute_error(y_test, predictions)
    rmse = np.sqrt(mean_squared_error(y_test, predictions))
    r2 = r2_score(y_test, predictions)

    print(f"\n{name}")
    print(f"MAE  = {mae:.3f}")
    print(f"RMSE = {rmse:.3f}")
    print(f"R²   = {r2:.3f}")

    return [name, mae, rmse, r2]

In [31]:
results = []

In [32]:
# Model 1: predicting G3 using all features (including G1 and G2)

X = df.drop('G3', axis=1)
y = df['G3']

# Convert categorical variables
X = pd.get_dummies(X, drop_first=True)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# Train model
model1 = LinearRegression()

model1.fit(X_train, y_train)

# Predictions
predictions1 = model1.predict(X_test)

# Evaluation
result1 = evaluate_model(
    "Model 1 (With G1 & G2)",
    y_test,
    predictions1
)

results.append(result1)


Model 1 (With G1 & G2)
MAE  = 1.647
RMSE = 2.378
R²   = 0.724


In [33]:
# Model 2: predicting G3 without G1 and G2, using only student characteristics

X = df.drop(['G1', 'G2', 'G3'], axis=1)
y = df['G3']

# Convert categorical variables
X = pd.get_dummies(X, drop_first=True)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

# Train model
model2 = LinearRegression()

model2.fit(X_train, y_train)

# Predictions
predictions2 = model2.predict(X_test)

# Evaluation
result2 = evaluate_model(
    "Model 2 (Without G1 & G2)",
    y_test,
    predictions2
)

results.append(result2)


Model 2 (Without G1 & G2)
MAE  = 3.395
RMSE = 4.196
R²   = 0.141


In [34]:
comparison = pd.DataFrame(
    results,
    columns=['Model', 'MAE', 'RMSE', 'R2']
)

comparison

,Model,MAE,RMSE,R2
0,Model 1 (With G1 & G2),1.646666,2.378370,0.724134
1,Model 2 (Without G1 & G2),3.395261,4.195681,0.141492


In [35]:
comparison.sort_values(by='MAE')

,Model,MAE,RMSE,R2
0,Model 1 (With G1 & G2),1.646666,2.378370,0.724134
1,Model 2 (Without G1 & G2),3.395261,4.195681,0.141492


# Conclusion

Two Linear Regression models were developed to predict student final grades (G3).

## Model 1 Results

- MAE = 1.6466
- RMSE = 2.3783
- R² = 0.7241

## Model 2 Results

- MAE = 3.3952
- RMSE = 4.1956
- R² = 0.1414

## Discussion

Model 1 achieved better predictive performance because it included previous academic grades (G1 and G2), which are highly correlated with final performance.

Model 2 demonstrated that demographic, family, lifestyle, and study-related factors can also contribute to predicting student performance, although with lower accuracy.

## Key Finding

Previous academic achievement is the strongest predictor of final student grades. However, factors such as study time, absences, parental education, and family support also provide useful predictive information.

## Future Work

Future studies may compare additional machine learning algorithms such as:

- Decision Tree Regression
- Random Forest Regression
- Gradient Boosting Regression

to determine whether non-linear models can improve prediction accuracy.